### ISSUE TITLE
recipe for target 'Release/libtensorflow.so' failed when installing `tfjs-node`

### ISSUE THREAD

**Issue by d4nyll** [tensorflow/tfjs#515]
#### TensorFlow.js version
`@tensorflow/tfjs-node` - `0.1.8`

#### Browser version
N/A

#### Linux version

```
$ uname -a
Linux hostname 4.15.0-23-generic #25-Ubuntu SMP Wed May 23 18:02:16 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04 LTS
Release:	18.04
Codename:	bionic
```

#### Describe the problem or feature request
Whenever I try to install `@tensorflow/tfjs-node`, `node-gyp` exits with code `1`.

#### Code to reproduce the bug / link to feature request

```
$ yarn add @tensorflow/tfjs @tensorflow/tfjs-node
yarn add v1.7.0
info No lockfile found.
[1/4] Resolving packages...
⠁ (node:32362) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
[2/4] Fetching packages...
[3/4] Linking dependencies...
warning " > @tensorflow/tfjs-node@0.1.8" has unmet peer dependency "@tensorflow/tfjs-core@~0.12.1".
[4/4] Building fresh packages...
[1/2] ⠠ protobufjs
error ~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments:
Directory: ~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@3.6.2
gyp info using node@10.6.0 | linux | x64
gyp info spawn /usr/bin/python2
gyp info spawn args [ '~/.nvm/versions/node/v10.6.0/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '~/.nvm/versions/node/v10.6.0/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '~/.node-gyp/10.6.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=~/.node-gyp/10.6.0',
gyp info spawn args   '-Dnode_gyp_dir=~/.nvm/versions/node/v10.6.0/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=~/.node-gyp/10.6.0/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory '~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node/build'
  ACTION binding_gyp_tfjs_binding_target_download_libtensorflow Release/libtensorflow.so
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 14.0M  100 14.0M    0     0   178k      0  0:01:20  0:01:20 --:--:--  173k

gzip: stdin: unexpected end of file
tar: Child returned status 1
tar: Error is not recoverable: exiting now
tfjs_binding.target.mk:13: recipe for target 'Release/libtensorflow.so' failed
make: Leaving directory '~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node/build'
make: *** [Release/libtensorflow.so] Error 2
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (~/.nvm/versions/node/v10.6.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:258:23)
gyp ERR! stack     at ChildProcess.emit (events.js:182:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12)
gyp ERR! System Linux 4.15.0-23-generic
gyp ERR! command "~/.nvm/versions/node/v10.6.0/bin/node" "~/.nvm/versions/node/v10.6.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd ~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node
gyp ERR! node -v v10.6.0
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok
```
I tried with the LTS (8.11.3) version of Node, and get the same error:
```
$ yarn add @tensorflow/tfjs @tensorflow/tfjs-node
yarn add v1.7.0
info No lockfile found.
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
warning " > @tensorflow/tfjs-node@0.1.8" has unmet peer dependency "@tensorflow/tfjs-core@~0.12.1".
[4/4] Building fresh packages...
[1/2] ⡀ protobufjs
error ~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments:
Directory: ~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@3.6.2
gyp info using node@8.11.3 | linux | x64
gyp http GET https://nodejs.org/download/release/v8.11.3/node-v8.11.3-headers.tar.gz
gyp http 200 https://nodejs.org/download/release/v8.11.3/node-v8.11.3-headers.tar.gz
gyp http GET https://nodejs.org/download/release/v8.11.3/SHASUMS256.txt
gyp http 200 https://nodejs.org/download/release/v8.11.3/SHASUMS256.txt
gyp info spawn /usr/bin/python2
gyp info spawn args [ '~/.nvm/versions/node/v8.11.3/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '~/.nvm/versions/node/v8.11.3/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '~/.node-gyp/8.11.3/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=~/.node-gyp/8.11.3',
gyp info spawn args   '-Dnode_gyp_dir=~/.nvm/versions/node/v8.11.3/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=~/.node-gyp/8.11.3/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory '~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node/build'
  ACTION binding_gyp_tfjs_binding_target_download_libtensorflow Release/libtensorflow.so
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 14.0M  100 14.0M    0     0   224k      0  0:01:03  0:01:03 --:--:--  360k

gzip: stdin: unexpected end of file
tar: Child returned status 1
tar: Error is not recoverable: exiting now
tfjs_binding.target.mk:13: recipe for target 'Release/libtensorflow.so' failed
make: Leaving directory '~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node/build'
make: *** [Release/libtensorflow.so] Error 2
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (~/.nvm/versions/node/v8.11.3/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:258:23)
gyp ERR! stack     at emitTwo (events.js:126:13)
gyp ERR! stack     at ChildProcess.emit (events.js:214:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:198:12)
gyp ERR! System Linux 4.15.0-23-generic
gyp ERR! command "~/.nvm/versions/node/v8.11.3/bin/node" "~/.nvm/versions/node/v8.11.3/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd ~/playground/tensorflowjs_tmp/node_modules/@tensorflow/tfjs-node
gyp ERR! node -v v8.11.3
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok
```

**Comment by nkreeger** [comment]
Can you try running this script?

https://github.com/tensorflow/tfjs-node/blob/master/scripts/download-libtensorflow.sh

If you clone tfjs-node:
```sh
$ ./scripts/download-libtensorflow.sh linux-cpu
```

And post the output here - thanks!

**Comment by d4nyll** [comment]
There you go!

```
$ ./tfjs-node/scripts/download-libtensorflow.sh linux-cpu
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 14.0M  100 14.0M    0     0  6473k      0  0:00:02  0:00:02 --:--:-- 6473k

gzip: stdin: unexpected end of file
tar: Child returned status 1
tar: Error is not recoverable: exiting now
```

**Comment by nkreeger** [comment]
@d4nyll can you download from this link? https://storage.googleapis.com/tf-builds/libtensorflow_r1_8_linux_cpu.tar.gz I wonder if you have a firewall. I was able to download that link from two of my machines.

**Comment by d4nyll** [comment]
@nkreeger Yes I am able to download the archive. I have tried three different connections - home, work and mobile / data.

**Comment by d4nyll** [comment]
After running the command, the files are downloaded and unarchived properly. But it throws an error. It seems that if I run the command again, it succeeds because it skips the `then` block in the last `if` statement.

```
$ ./tfjs-node/scripts/download-libtensorflow.sh linux-cpu
deps/tensorflow/
deps/tensorflow/lib/libtensorflow.so
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 14.0M  100 14.0M    0     0  7325k      0  0:00:01  0:00:01 --:--:-- 7321k

gzip: stdin: unexpected end of file
tar: Child returned status 1
tar: Error is not recoverable: exiting now
d4nyll:~/playground$ ./tfjs-node/scripts/download-libtensorflow.sh linux-cpu
d4nyll:~/playground$
```

**Comment by d4nyll** [comment]
I think I figured it out. I had a `~/.curlrc` file with the line:

    -w "\n"

This set the `--write-out` option to a newline character `\n`, which causes cURL to append a `\n` character to the end of the output. This is why the file _appeared_ to be corrupted.

**Comment by d4nyll** [comment]
I think having a `.curlrc` where the `--write-out` option is specified is quite common. Here are several Stack* questions related to this issue:

* [Automatically add newline at end of curl response body](https://stackoverflow.com/q/12849584/3966682)
* [Append line break to command output](https://superuser.com/q/594883/407572)

We can either use the temporary file option, or override the `-w` option. I think the latter is more 'elegant' and have created a [PR](https://github.com/tensorflow/tfjs-node/pull/121) to solve this issue.

**Comment by nkreeger** [comment]
@d4nyll thanks for the fix! I'll get a new build out early next week.

**Comment by nkreeger** [comment]
Actually - just published 0.1.9 since we had one small fix queued.

**Comment by d4nyll** [comment]
@nkreeger Thanks for your quick response. Just to confirm, I am now using `tfjs-node` `v0.1.9` and it's working like a charm!

```
$ yarn add @tensorflow/tfjs @tensorflow/tfjs-node
yarn add v1.7.0
info No lockfile found.
(node:29358) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
warning " > @tensorflow/tfjs-node@0.1.9" has unmet peer dependency "@tensorflow/tfjs-core@~0.12.1".
[4/4] Building fresh packages...
success Saved lockfile.
success Saved 23 new dependencies.
info Direct dependencies
├─ @tensorflow/tfjs-node@0.1.9
└─ @tensorflow/tfjs@0.12.2
info All dependencies
├─ @protobufjs/aspromise@1.1.2
├─ @protobufjs/base64@1.1.2
├─ @protobufjs/codegen@2.0.4
├─ @protobufjs/eventemitter@1.1.0
├─ @protobufjs/fetch@1.1.0
├─ @protobufjs/float@1.0.2
├─ @protobufjs/path@1.1.2
├─ @protobufjs/pool@1.1.0
├─ @protobufjs/utf8@1.1.0
├─ @tensorflow/tfjs-converter@0.5.1
├─ @tensorflow/tfjs-core@0.12.4
├─ @tensorflow/tfjs-layers@0.7.0
├─ @tensorflow/tfjs-node@0.1.9
├─ @tensorflow/tfjs@0.12.2
├─ @types/long@3.0.32
├─ @types/node@8.10.21
├─ bindings@1.3.0
├─ long@4.0.0
├─ protobufjs@6.8.6
├─ punycode@1.3.2
├─ querystring@0.2.0
├─ seedrandom@2.4.3
└─ url@0.11.0
Done in 11.69s.
```
